Integrate `util/groupcommit` package with the coordinator table commit (#1728)
scalar-labs/scalardb: docs/api-guide.md

<div class="notice--info">{{ notice--info | markdownify }}</div>

### Group commit for the Coordinator table

> The Coordinator table used for Consensus Commit transactions is a vital data store, and it is recommended to use robust storage for it. However, utilizing more robust storage options, such as internally leveraging multi-AZ or multi-region replication, may lead to increased latency when writing records to the storage, resulting in poor throughput performance. ScalarDB provides a group commit feature for the Coordinator table that groups multiple record writes into a single write operation, improving write throughput. However, it may increase latency.

Author: Good point. Basically group commit tends to increase latency as it adds an additional step to wait multiple requests. On the other hand, it may reduces workload on the database by reducing request, resulting in reducing the latency. So, we can say it may depends on the situation.

> The Coordinator table used for Consensus Commit transactions is a vital data store, and it is recommended to use robust storage for it. However, utilizing more robust storage options, such as internally leveraging multi-AZ or multi-region replication, may lead to increased latency when writing records to the storage, resulting in poor throughput performance. ScalarDB provides a group commit feature for the Coordinator table that groups multiple record writes into a single write operation, improving write throughput. However, it may increase latency.

Reviewer: I see. How about the following then, which would make it a little more generic?

```suggestion
The Coordinator table that is used for Consensus Commit transactions is a vital data store, and using robust storage for it is recommended. However, utilizing more robust storage options, such as internally leveraging multi-AZ or multi-region replication, may lead to increased latency when writing records to the storage, resulting in poor throughput performance. 

ScalarDB provides a group commit feature for the Coordinator table that groups multiple record writes into a single write operation, improving write throughput. In this case, latency may increase or decrease, depending on your environment.
```

> The Coordinator table used for Consensus Commit transactions is a vital data store, and it is recommended to use robust storage for it. However, utilizing more robust storage options, such as internally leveraging multi-AZ or multi-region replication, may lead to increased latency when writing records to the storage, resulting in poor throughput performance. ScalarDB provides a group commit feature for the Coordinator table that groups multiple record writes into a single write operation, improving write throughput. However, it may increase latency.

Author: > depending on your environment

This sounded a bit too broad scope. How about "depending on the underlying database (or storage) and the workload" or something?

> The Coordinator table used for Consensus Commit transactions is a vital data store, and it is recommended to use robust storage for it. However, utilizing more robust storage options, such as internally leveraging multi-AZ or multi-region replication, may lead to increased latency when writing records to the storage, resulting in poor throughput performance. ScalarDB provides a group commit feature for the Coordinator table that groups multiple record writes into a single write operation, improving write throughput. However, it may increase latency.

Reviewer: Thank you for clarifying! I wasn't sure what the scope was, so your suggestion sounds good to me!

```suggestion
The Coordinator table that is used for Consensus Commit transactions is a vital data store, and using robust storage for it is recommended. However, utilizing more robust storage options, such as internally leveraging multi-AZ or multi-region replication, may lead to increased latency when writing records to the storage, resulting in poor throughput performance.

ScalarDB provides a group commit feature for the Coordinator table that groups multiple record writes into a single write operation, improving write throughput. In this case, latency may increase or decrease, depending on the underlying database and the workload.
```

> The Coordinator table used for Consensus Commit transactions is a vital data store, and it is recommended to use robust storage for it. However, utilizing more robust storage options, such as internally leveraging multi-AZ or multi-region replication, may lead to increased latency when writing records to the storage, resulting in poor throughput performance. ScalarDB provides a group commit feature for the Coordinator table that groups multiple record writes into a single write operation, improving write throughput. However, it may increase latency.

Author: @josh-wong Thank you!

##### Custom transaction ID passed by users

The group commit feature implicitly generates an internal value and uses it as a part of transaction ID. Therefore, a custom transaction ID manually passed by users via `com.scalar.db.transaction.consensuscommit.ConsensusCommitManager.begin(String txId)` or `com.scalar.db.transaction.consensuscommit.TwoPhaseConsensusCommitManager.begin(String txId)` can't be used as-is for later API calls. You need to use a transaction ID returned from`com.scalar.db.transaction.consensuscommit.ConsensusCommit.getId()` or `com.scalar.db.transaction.consensuscommit.TwoPhaseConsensusCommit.getId()` instead.

```java
   // This custom transaction ID needs to be used for ScalarDB transactions for some reason...
   String myTxId = UUID.randomUUID().toString();

   ...

   DistributedTransaction transaction = manager.begin(myTxId);

   ...

   // When the group commit feature is enabled, a custom transaction ID passed by users can't be used as is.
   // logger.info("The transaction state: {}", manager.getState(myTxId));
   logger.info("The transaction state: {}", manager.getState(transaction.getId()));
```

##### Commit in Two-Phase Commit Interface

> The group commit feature manages all in-flight transactions in memory. When the feature is enabled with the two-phase commit interface, the information must be solely maintained by the coordinator service to prevent conflicts caused by participant services' inconsistent writes to the coordinator table, which may contain different transaction distributions over groups. Consequently, `com.scalar.db.transaction.consensuscommit.TwoPhaseConsensusCommit.commit()` must be called first by the coordinator service before being called by participant services.

Reviewer: Is there a more accurate word we can use instead of `in-flight`, like `pending` or `in-progress` be accurate?

```suggestion
The group commit feature manages all in-flight transactions in memory. When the feature is enabled with the two-phase commit interface, the information must be solely maintained by the coordinator service to prevent conflicts caused by participant services' inconsistent writes to the Coordinator table, which may contain different transaction distributions over groups. Consequently, `com.scalar.db.transaction.consensuscommit.TwoPhaseConsensusCommit.commit()` must be called first by the coordinator service before being called by participant services.
```

Author: `ongoing transactions` might be better?

Reviewer: `ongoing transactions` sounds good to me! Thanks!

<div class="notice--info">{{ notice--info | markdownify }}</div>

### Group commit for the Coordinator table

The Coordinator table that is used for Consensus Commit transactions is a vital data store, and using robust storage for it is recommended. However, utilizing more robust storage options, such as internally leveraging multi-AZ or multi-region replication, may lead to increased latency when writing records to the storage, resulting in poor throughput performance.

ScalarDB provides a group commit feature for the Coordinator table that groups multiple record writes into a single write operation, improving write throughput. In this case, latency may increase or decrease, depending on the underlying database and the workload.

> To enable the group commit feature, add the following configuration:
> ```properties

Reviewer: ````suggestion
To enable the group commit feature, add the following configuration:

```properties
````

<div class="notice--info">{{ notice--info | markdownify }}</div>

### Group commit for the Coordinator table

The Coordinator table that is used for Consensus Commit transactions is a vital data store, and using robust storage for it is recommended. However, utilizing more robust storage options, such as internally leveraging multi-AZ or multi-region replication, may lead to increased latency when writing records to the storage, resulting in poor throughput performance.

ScalarDB provides a group commit feature for the Coordinator table that groups multiple record writes into a single write operation, improving write throughput. In this case, latency may increase or decrease, depending on the underlying database and the workload.

To enable the group commit feature, add the following configuration:

> 
> #### Configurations
> 

Reviewer: I don't think we need a heading here since we introduce the code block in the previous sentence (`To enable the group commit feature, add the following configuration:`).

```suggestion

```

```properties
# By default, this configuration is set to `false`.
scalar.db.consensus_commit.coordinator.group_commit.enabled=true

# These properties are for tuning the performance of the group commit feature.
# scalar.db.consensus_commit.coordinator.group_commit.group_size_fix_timeout_millis=40
# scalar.db.consensus_commit.coordinator.group_commit.delayed_slot_move_timeout_millis=800
# scalar.db.consensus_commit.coordinator.group_commit.old_group_abort_timeout_millis=30000
# scalar.db.consensus_commit.coordinator.group_commit.timeout_check_interval_millis=10
# scalar.db.consensus_commit.coordinator.group_commit.metrics_monitor_log_enabled=true
```

> #### Limitations
> 
> ##### Custom transaction ID passed by users

Reviewer: So that we're not stacking headings without text between them, I've added the following sentence. Please feel free to revise it as necessary.

```suggestion
#### Limitations

This section describes the limitations of the group commit feature.

##### Custom transaction ID passed by users
```

The group commit feature implicitly generates an internal value and uses it as a part of transaction ID. Therefore, a custom transaction ID manually passed by users via `com.scalar.db.transaction.consensuscommit.ConsensusCommitManager.begin(String txId)` or `com.scalar.db.transaction.consensuscommit.TwoPhaseConsensusCommitManager.begin(String txId)` can't be used as is for later API calls. You need to use a transaction ID returned from`com.scalar.db.transaction.consensuscommit.ConsensusCommit.getId()` or `com.scalar.db.transaction.consensuscommit.TwoPhaseConsensusCommit.getId()` instead.

```java
   // This custom transaction ID needs to be used for ScalarDB transactions.
   String myTxId = UUID.randomUUID().toString();

   ...

   DistributedTransaction transaction = manager.begin(myTxId);

   ...

   // When the group commit feature is enabled, a custom transaction ID passed by users can't be used as is.
   // logger.info("The transaction state: {}", manager.getState(myTxId));
   logger.info("The transaction state: {}", manager.getState(transaction.getId()));
```

> ##### Commit in Two-Phase Commit Interface

Reviewer: ```suggestion
##### Commit in two-phase commit interface
```
